src/home/homed-bus.c

```diff
@@ -64,3 +68,72 @@ int bus_message_read_home_record(sd_bus_message *m, UserRecordLoadFlags flags, U
         *ret = TAKE_PTR(hr);
         return 0;
 }
+
+int bus_message_read_blobs(sd_bus_message *m, Hashmap **ret, sd_bus_error *error) {
+        _cleanup_hashmap_free_ Hashmap *blobs = NULL;
+        int r;
+
+        assert(m);
+        assert(ret);
+
+        /* We want to differentiate between blobs being NULL (not passed at all)
+         * and empty (passed from dbus, but it was empty) */
+        r = hashmap_ensure_allocated(&blobs, &blob_fd_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_enter_container(m, 'a', "{sh}");
+        if (r < 0)
+                return r;
+
+        for (;;) {
+                _cleanup_free_ char *filename = NULL;
+                _cleanup_close_ int fd = -EBADF;
+                const char *_filename = NULL;
+                int _fd, flags;
+
+                r = sd_bus_message_read(m, "{sh}", &_filename, &_fd);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                filename = strdup(_filename);
+                if (!filename)
+                        return -ENOMEM;
+
+                fd = fcntl(_fd, F_DUPFD_CLOEXEC, 3);
+                if (fd < 0)
+                        return -errno;
+
+                r = suitable_blob_filename(filename);
+                if (r < 0)
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid blob directory filename: %s", filename);
+
+                r = fd_verify_regular(fd);
+                if (r < 0)
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "FD for %s is not a regular file", filename);
+
+                flags = fcntl(fd, F_GETFL);
+                if (flags < 0)
+                        return -errno;
+
+                /* Refuse fds w/ unexpected flags set. In particular, we don't want to permit O_PATH FDs, since
+                 * those don't actually guarentee that the client has access to the file. */
+                if ((flags & ~(O_ACCMODE|RAW_O_LARGEFILE)) != 0)
```

Comment:
So apparently `O_NOFOLLOW` continues existing in the list of flags after `open`.

So if a client opens the file with `O_NOFOLLOW`, then homed will refuse the FD on the grounds that `O_NOFOLLOW` is not one of the allowlisted flags here.

Should we add `O_NOFOLLOW` to the list here? What about the other flags (`O_NOATIME`, `O_DIRECT`, `O_DYSNC`, `FASYNC`)?

Reply:
hmm, are sure that O_NOFOLLOW is kept around? i am very surprised by this.

I'd be careful with O_DIRECT and O_DSYNC. Rather refuse too much than too little.

it sounds to me as if we should refuse everything remotely questionable, as long as noone complains.

Reply:
(Looks like I caught GitHub deleting another one of my comments... I opened a support ticket with them; losing the integrity of the conversation is quite bad...)

well that's the behavior I observed. I even printed out the offending flags and O_NOFOLLOW and compared against the definitions of the various O_* flags, and O_NOFOLLOW was the one that was set and caused the check to fail. Removing O_NOFOLLOW in AccountsService fixed the error I was getting from homed too.

No idea if it's intentional 🤷